File: go/src/RumbleRacing/core/asset/txf/TXF.go
package txf

import (
	"encoding/binary"
	"errors"
	"fmt"
	"rumble-reader/chunk/shoc"
)

type TXF struct {
	rawData []byte

	resourceName string
	shocHeader   shoc.SHDR

	header         *HEAD
	textureHeaders []*ZTHE
	clutHeader     *CLHE
	textureData    *TXDA
	clutData       *CLDA
}

func ParseTXF(buf []byte, hdr shoc.SHDR, resName string) (*TXF, error) {
	txfAsset := TXF{
		rawData:      buf,
		shocHeader:   hdr,
		resourceName: resName,
	}

	chunks, err := splitTaggedChunks(buf[8:])

	if err != nil {
		panic(err)
	}

	for _, chunk := range chunks {
		// fmt.Println(chunkIndex, len(chunk), string(chunk[0:4]))
		tag := string(chunk[0:4])

		switch tag {
		case "HEAD":
			{
				if head, err := parseHEAD(chunk); err == nil {
					if txfAsset.header != nil {
						return nil, errors.New("multiple HEAD in TXF file")
					}
					txfAsset.header = head
				}
			}
		case "ZTHE":
			{
				if zthe, err := parseZTHE(chunk); err == nil {
					txfAsset.textureHeaders = append(txfAsset.textureHeaders, zthe)
				}
			}
		case "CLHE":
			if clhe, err := parseCLHE(chunk); err == nil {
				if txfAsset.clutHeader != nil {
					return nil, errors.New("multiple CLHE in TXF file")
				}
				txfAsset.clutHeader = clhe
			}
		case "TXDA":
			if txda, err := parseTXDA(chunk); err == nil {
				if txfAsset.textureData != nil {
					return nil, errors.New("multiple TXDA in TXF file")
				}
				txfAsset.textureData = txda
			}
		case "CLDA":
			if clda, err := parseCLDA(chunk); err == nil {
				if txfAsset.clutData != nil {
					return nil, errors.New("multiple CLDA in TXF file")
				}
				txfAsset.clutData = clda
			}
		default:
			{
				panic("Unknown TXF chunk tag: " + tag)
			}
		}
	}

	if int(txfAsset.header.CLHEIterations) != len(txfAsset.clutHeader.Entries) {
		// panic("txf header clheIterations is not equal to actual entries in CLUT header")
	}

	return &txfAsset, nil
}

func splitTaggedChunks(buf []byte) ([][]byte, error) {
	var chunks [][]byte
	offset := 0

	for offset+8 <= len(buf) {
		tag := buf[offset : offset+4]
		// fmt.Println(string(tag))
		size := binary.LittleEndian.Uint32(buf[offset+4 : offset+8])
		offset += 8

		if offset+int(size) > len(buf) {
			return nil, fmt.Errorf("invalid size %d at offset %d", size, offset)
		}

		data := buf[offset : offset+int(size)]
		offset += int(size)

		chunk := make([]byte, 0, 8+len(data))
		chunk = append(chunk, tag...)
		chunk = append(chunk,
			buf[offset-int(size)-4:offset-int(size)]...,
		)
		chunk = append(chunk, data...)

		chunks = append(chunks, chunk)
	}

	return chunks, nil
}

func (g *TXF) GetType() string {
	return "txf"
}

func (g *TXF) RawData() []byte {
	return g.rawData
}

func (g *TXF) Header() shoc.SHDR {
	return g.shocHeader
}

File: src/RumbleRacing/asset/txf/TXF.ts
import { SHDR } from "../../chunk/shoc/shdr";
import ArrayBufferSlice from "../../../ArrayBufferSlice";
import { readString } from "../../../util";
import { HEAD, parseHEAD } from "./HEAD";
import { ZTHE, parseZTHE } from "./ZTHE";
import { CLHE, parseCLHE } from "./CLHE";
import { TXDA, parseTXDA } from "./TXDA";
import { CLDA, parseCLDA } from "./CLDA";
import { Texture, extractTexturesFromZTHE } from "./texture";

export interface TXF {
  kind: "TXF";
  rawData: ArrayBufferSlice;
  resourceName: string;
  shocHeader: SHDR;
  header: HEAD;
  textureHeaders: ZTHE[];
  clutHeader: CLHE;
  textureData: TXDA;
  clutData: CLDA;
}

function splitTaggedChunks(buf: ArrayBufferSlice): ArrayBufferSlice[] {
  const view = buf.createDataView();
  const chunks: ArrayBufferSlice[] = [];
  let offset = 0;

  while (offset + 8 <= buf.byteLength) {
    const size = view.getUint32(offset + 4, true);

    if (offset + 8 + size > buf.byteLength) {
      throw new Error(`invalid size ${size} at offset ${offset + 8}`);
    }

    chunks.push(buf.subarray(offset, 8 + size));
    offset += 8 + size;
  }

  return chunks;
}

export function parseTXF(
  buf: ArrayBufferSlice,
  hdr: SHDR,
  resName: string,
): TXF {
  const chunks = splitTaggedChunks(buf.subarray(8));

  let header: HEAD | null = null;
  const textureHeaders: ZTHE[] = [];
  let clutHeader: CLHE | null = null;
  let textureData: TXDA | null = null;
  let clutData: CLDA | null = null;

  for (const chunk of chunks) {
    const tag = readString(chunk, 0, 4, false);

    switch (tag) {
      case "HEAD":
        if (header !== null) throw new Error("multiple HEAD in TXF file");
        header = parseHEAD(chunk);
        break;
      case "ZTHE":
        textureHeaders.push(parseZTHE(chunk));
        break;
      case "CLHE":
        if (clutHeader !== null) throw new Error("multiple CLHE in TXF file");
        clutHeader = parseCLHE(chunk);
        break;
      case "TXDA":
        if (textureData !== null) throw new Error("multiple TXDA in TXF file");
        textureData = parseTXDA(chunk);
        break;
      case "CLDA":
        if (clutData !== null) throw new Error("multiple CLDA in TXF file");
        clutData = parseCLDA(chunk);
        break;
      default:
        throw new Error("Unknown TXF chunk tag: " + tag);
    }
  }

  if (!header || !clutHeader || !textureData || !clutData) {
    throw new Error("TXF missing required chunks");
  }

  return {
    kind: "TXF",
    rawData: buf,
    resourceName: resName,
    shocHeader: hdr,
    header,
    textureHeaders,
    clutHeader,
    textureData,
    clutData,
  };
}

export function getTextures(txf: TXF): Texture[] {
  const textures: Texture[] = [];

  for (let i = 0; i < txf.textureHeaders.length; i++) {
    const zthe = txf.textureHeaders[i];
    for (let j = 0; j < zthe.textures.length; j++) {
      const tex = zthe.textures[j];
      const clhe = txf.clutHeader.entries[tex.clutHeaderIndex];
      const extracted = extractTexturesFromZTHE(txf, clhe, tex);
      textures.push(...extracted);
    }
  }

  return textures;
}
